### ISSUE TITLE
Fastest way to read all items with a single partition key

### ISSUE THREAD

**Issue by simonziegler** [Azure/azure-cosmos-dotnet-v3#2014]
**Is your feature request related to a problem? Please describe.**
I have a container where my data is organized by partition key such that my most common reads return all items with a matching partition key. Stress testing the database in a development environment I observe the following metrics:

- Once partition holds 200 documents totalling 155Kb. Using a V3 feed iterator from a call to `GetItemLinqQueryable` where only the partition key is set takes circa 80ms round trip.
- A smaller partition holds 46 documents totalling 34Kb. The same code has a 25ms query round trip. So it seems that latency is no more than 25ms.
- My cosmos db service is a singleton and I am using Direct mode.

**Describe the solution you'd like**
I would like to know the most performant way to specify a partition key and have the entire dataset returned to me. Note that I am using a custom polymorphic Newtonsoft serializer that takes less than 3ms to serialize the larger dataset (on my PC of course)

**Describe alternatives you've considered**
None have occurred to me yet.

**Additional context**
- I am just outside London, UK and the region is UK South.
- Here is the debugging output from the two queries (longer first, smaller and faster second):

```json
 {
  "DiagnosticVersion": "2",
  "Summary": {
    "StartUtc": "2020-11-18T11:23:09.6599136Z",
    "TotalElapsedTimeInMs": 77.4413,
    "UserAgent": "cosmos-netstandard-sdk/3.15.1|3.14.1|03|X64|Microsoft Windows 10.0.19042|.NET 5.0.0|",
    "TotalRequestCount": 1,
    "FailedRequestCount": 0
  },
  "Context": [
    {
      "Id": "InitializeContainerResourceId",
      "ElapsedTimeInMs": 0.0197
    },
    {
      "Id": "InitializeContinuation",
      "ElapsedTimeInMs": 0.0183
    },
    {
      "Id": "Microsoft.Azure.Cosmos.Handlers.DiagnosticsHandler",
      "HandlerElapsedTimeInMs": 77.3049
    },
    {
      "Id": "SystemInfo",
      "CpuHistory": "(2020-11-18T11:07:04.2767057Z 0.000), (2020-11-18T11:07:14.8560834Z 12.001), (2020-11-18T11:07:58.9552006Z 8.236), (2020-11-18T11:22:44.2776025Z 12.500), (2020-11-18T11:22:54.2813467Z 6.250), (2020-11-18T11:23:04.2689887Z 0.000)"
    },
    {
      "Id": "Microsoft.Azure.Cosmos.Handlers.RetryHandler",
      "HandlerElapsedTimeInMs": 77.2922
    },
    {
      "Id": "Microsoft.Azure.Cosmos.Handlers.RouterHandler",
      "HandlerElapsedTimeInMs": 77.2688
    },
    {
      "Id": "Microsoft.Azure.Cosmos.Handlers.TransportHandler",
      "HandlerElapsedTimeInMs": 1.9519
    },
    {
      "Id": "AggregatedClientSideRequestStatistics",
      "ContactedReplicas": [],
      "RegionsContacted": [
        "https://xxxx.azure.com:443/"
      ],
      "FailedReplicas": []
    },
    {
      "Id": "Microsoft.Azure.Documents.ServerStoreModel",
      "ElapsedTimeInMs": 77.2095
    },
    {
      "Id": "StoreResponseStatistics",
      "StartTimeUtc": "2020-11-18T11:23:09.6600784Z",
      "ResponseTimeUtc": "2020-11-18T11:23:09.737163Z",
      "ElapsedTimeInMs": 77.0846,
      "ResourceType": "Document",
      "OperationType": "ReadFeed",
      "LocationEndpoint": "https://xxxx.azure.com:443/",
      "ActivityId": "43c04d39-3789-4ce1-86fe-c0ab863e3120",
      "StoreResult": "StorePhysicalAddress: rntbd://cdb-ms-prod-uksouth1-fd11.documents.azure.com:14073/apps/9f3cfc73-2966-4a41-a19d-adcf3216f2ef/services/94a6c9c7-7887-4615-b51a-43fdaae925de/partitions/111101ba-b38c-4c16-bda6-b6f0243769c6/replicas/132475379554821069s/, LSN: 408704, GlobalCommittedLsn: 408703, PartitionKeyRangeId: 0, IsValid: True, StatusCode: 200, SubStatusCode: 0, RequestCharge: 16.05, ItemLSN: -1, SessionToken: -1#408704, UsingLocalLSN: True, TransportException: null"
    }
  ]
}
```

```json
{
  "DiagnosticVersion": "2",
  "Summary": {
    "StartUtc": "2020-11-18T11:23:09.1205241Z",
    "TotalElapsedTimeInMs": 22.8376,
    "UserAgent": "cosmos-netstandard-sdk/3.15.1|3.14.1|03|X64|Microsoft Windows 10.0.19042|.NET 5.0.0|",
    "TotalRequestCount": 1,
    "FailedRequestCount": 0
  },
  "Context": [
    {
      "Id": "InitializeContainerResourceId",
      "ElapsedTimeInMs": 0.1936
    },
    {
      "Id": "InitializeContinuation",
      "ElapsedTimeInMs": 0.1024
    },
    {
      "Id": "Microsoft.Azure.Cosmos.Handlers.DiagnosticsHandler",
      "HandlerElapsedTimeInMs": 22.3086
    },
    {
      "Id": "SystemInfo",
      "CpuHistory": "(2020-11-18T11:07:04.2767057Z 0.000), (2020-11-18T11:07:14.8560834Z 12.001), (2020-11-18T11:07:58.9552006Z 8.236), (2020-11-18T11:22:44.2776025Z 12.500), (2020-11-18T11:22:54.2813467Z 6.250), (2020-11-18T11:23:04.2689887Z 0.000)"
    },
    {
      "Id": "Microsoft.Azure.Cosmos.Handlers.RetryHandler",
      "HandlerElapsedTimeInMs": 22.2839
    },
    {
      "Id": "Microsoft.Azure.Cosmos.Handlers.RouterHandler",
      "HandlerElapsedTimeInMs": 22.232
    },
    {
      "Id": "Microsoft.Azure.Cosmos.Handlers.TransportHandler",
      "HandlerElapsedTimeInMs": 2.5945
    },
    {
      "Id": "AggregatedClientSideRequestStatistics",
      "ContactedReplicas": [],
      "RegionsContacted": [
        "https://xxxx.azure.com:443/"
      ],
      "FailedReplicas": []
    },
    {
      "Id": "Microsoft.Azure.Documents.ServerStoreModel",
      "ElapsedTimeInMs": 22.0594
    },
    {
      "Id": "StoreResponseStatistics",
      "StartTimeUtc": "2020-11-18T11:23:09.1211848Z",
      "ResponseTimeUtc": "2020-11-18T11:23:09.1430213Z",
      "ElapsedTimeInMs": 21.8365,
      "ResourceType": "Document",
      "OperationType": "ReadFeed",
      "LocationEndpoint": "https://xxxx.azure.com:443/",
      "ActivityId": "f21de951-9407-42a2-8ee9-246fc398c017",
      "StoreResult": "StorePhysicalAddress: rntbd://cdb-ms-prod-uksouth1-fd11.documents.azure.com:14073/apps/9f3cfc73-2966-4a41-a19d-adcf3216f2ef/services/94a6c9c7-7887-4615-b51a-43fdaae925de/partitions/111101ba-b38c-4c16-bda6-b6f0243769c6/replicas/132475379554821069s/, LSN: 408704, GlobalCommittedLsn: 408703, PartitionKeyRangeId: 0, IsValid: True, StatusCode: 200, SubStatusCode: 0, RequestCharge: 4.59, ItemLSN: -1, SessionToken: -1#408704, UsingLocalLSN: True, TransportException: null"
    }
  ]
}
```

**Comment by j82w** [comment]
Please make sure you are following the [performance recommendations](https://docs.microsoft.com/en-us/azure/cosmos-db/performance-tips-dotnet-sdk-v3-sql) for the SDK. You are already following a few the more important suggestions like singleton and direct mode.

@bchong95 or @timsander1 can you please answer the above question.

**Comment by timsander1** [comment]
This sounds like reasonable latency between your local machine in the UK and UK South. Querying for data with an equality filter on the partition key will filter data to that specific partition.

An alternative to querying data is to use the change feed pull model to read all the data within a partition key: https://docs.microsoft.com/azure/cosmos-db/change-feed-pull-model#consuming-a-partition-keys-changes

I think an in-partition query (like you are already doing) is most appropriate for this scenario, though.

**Comment by simonziegler** [comment]
Thanks. I set the partition key in the feed options. I haven't heard of the change feed pull model so I'll take a look at that. The thing that slightly puzzles me is the time taken to get through the feed iterator as the dataset grows. This is out of proportion with any RU limits, and the same query executed directly on the azure portal executes within a handful of milliseconds, so I'm trying to work out what causes the delay. So it could be something I'm doing wrong (seems not, especially because I designed the data model to be PK centric this way), latency which I doubt because the feed iterator only executes the first loop or something else.

Basically I know that this dataset has the potential to grow tenfold, at which point I anticipate the delay at that stage being meaningful for my users. I'll probably have a trial dataset in a few days to test this.

Any further thoughts much appreciated!

**Comment by timsander1** [comment]
If you are interesting in reading a substantial amount of data from a particular partition key (50 MB or more), then change feed pull model is preferred. 

As your query has more results, the RU charge will increase. This is simply because we have to load more documents and there isn't a way to avoid this. You may find that reading using the change feed pull model is more efficient and faster than performing queries.

**Comment by simonziegler** [comment]
Thanks @timsander1. I know that the RU will be whatever it needs to be and that doesn't bother me in the slightest. Realistically I expect a maximum real world single-partition dataset to be circa 1MB to 2MB. What I'm aiming for is a pace of read access that seems either imperceptible or negligible to my users. Imperceptible will be anything under 200ms (I have some client side animations that give me a timing head start), and negligible is under 500ms. Over that I get a bit twitchy, and I see things breaching that limit as it goes right now. I clearly have some experimentation to do!

**Comment by timsander1** [comment]
Achieving < 200 ms latency is totally reasonable with an in-partition query. Just make sure you are testing from a VM in the same Azure region as Cosmos DB

**Comment by simonziegler** [comment]
That's excellent. I'll add some logging to my app for its Azure deployment. A dumb question: when you refer to latency, do you specifically mean network latency or the total round trip of a query of the magnitude I'm expecting (so between say 1k and 2k documents and 1MB and 2MB combined document size)? Thanks!

**Comment by timsander1** [comment]
Both should be under 200ms

**Comment by simonziegler** [comment]
Excellent to know, so now I have a great handle on whether future performance is right, and if not I can go hunting for reasons in my code. I really appreciate your help on this and that you responded so swiftly!
